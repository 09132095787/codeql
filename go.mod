module github.com/github/codeql-go

go 1.14

require (
	github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d
	github.com/Masterminds/squirrel v1.4.0
	github.com/lib/pq v1.8.0
	golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
	golang.org/x/mod v0.3.0
	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
	golang.org/x/tools v0.0.0-20200827010519-17fd2f27a9e3
)
