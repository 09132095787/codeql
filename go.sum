cloud.google.com/go v0.34.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d h1:W1diKnDQkXxNDhghdBSbQ4LI/E1aJNTwpqPp3KtlB8w=
github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d/go.mod h1:nuWgzSkT5PnyOd+272uUmV0dnAnAn42Mk7PiQC5VzN4=
github.com/Masterminds/squirrel v1.4.0 h1:he5i/EXixZxrBUWcxzDYMiju9WZ3ld/l7QBNuo/eN3w=
github.com/Masterminds/squirrel v1.4.0/go.mod h1:yaPeOnPG5ZRwL9oKdTsO/prlkPbXWZlRVMQ/gGlzIuA=
github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
github.com/golang/protobuf v1.2.0 h1:P3YflyNX/ehuJFLhxviNdFxQPkGK5cDcApsge1SqnvM=
github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
github.com/lann/builder v0.0.0-20180802200727-47ae307949d0 h1:SOEGU9fKiNWd/HOJuq6+3iTQz8KNCLtVX6idSoTLdUw=
github.com/lann/builder v0.0.0-20180802200727-47ae307949d0/go.mod h1:dXGbAdH5GtBTC4WfIxhKZfyBF/HBFgRZSWwZ9g/He9o=
github.com/lann/ps v0.0.0-20150810152359-62de8c46ede0 h1:P6pPBnrTSX3DEVR4fDembhRWSsG5rVo6hYhAB/ADZrk=
github.com/lann/ps v0.0.0-20150810152359-62de8c46ede0/go.mod h1:vmVJ0l/dxyfGW6FmdpVm2joNMFikkuWg0EoCKLGUMNw=
github.com/lib/pq v1.8.0 h1:9xohqzkUwzR4Ga4ivdTcawVS89YSDVxXMa3xJX3cGzg=
github.com/lib/pq v1.8.0/go.mod h1:AlVN5x4E4T544tWzH6hKfbfQvm3HdbOxrmggDNAPY9o=
github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
github.com/stretchr/testify v1.2.2/go.mod h1:a8OnRcib4nhh0OaRAV+Yts87kKdq0PP7pXfy6kDkUVs=
github.com/yuin/goldmark v1.1.32/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9decYSb74=
golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2/go.mod h1:djNgcEr1/C05ACkg1iLfiJU5Ep61QUkGW8qpdssI0+w=
golang.org/x/crypto v0.0.0-20191011191535-87dc89f01550/go.mod h1:yigFU9vqHzYiE8UmvKecakEJjdnWj3jj499lnFckfCI=
golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9 h1:psW17arqaxU48Z5kZ0CQnkZWQJsqcURM6tKiBApRjXI=
golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
golang.org/x/mod v0.1.1-0.20191105210325-c90efee705ee/go.mod h1:QqPTAvyqsEbceGzBzNggFXnrqF1CaUcvgkdR5Ot7KZg=
golang.org/x/mod v0.2.0 h1:KU7oHjnv3XNWfa5COkzUifxZmxp1TyI7ImMXqFxLwvQ=
golang.org/x/mod v0.2.0/go.mod h1:s0Qsj1ACt9ePp/hMypM3fl4fZqREWJwdYDEqhRiZZUA=
golang.org/x/mod v0.3.0 h1:RM4zey1++hCTbCVQfnWeKs9/IEsaBLA8vTkd0WVtmH4=
golang.org/x/mod v0.3.0/go.mod h1:s0Qsj1ACt9ePp/hMypM3fl4fZqREWJwdYDEqhRiZZUA=
golang.org/x/net v0.0.0-20180724234803-3673e40ba225/go.mod h1:mL1N/T3taQHkDXs73rZJwtUhF3w3ftmwwsq0BUmARs4=
golang.org/x/net v0.0.0-20190108225652-1e06a53dbb7e/go.mod h1:mL1N/T3taQHkDXs73rZJwtUhF3w3ftmwwsq0BUmARs4=
golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3/go.mod h1:t9HGtf8HONx5eT2rtn7q6eTqICYqUVnKs3thJo3Qplg=
golang.org/x/net v0.0.0-20190620200207-3b0461eec859/go.mod h1:z5CRVTTTmAJ677TzLLGU+0bjPO0LkuOLi4/5GtJWs/s=
golang.org/x/net v0.0.0-20200625001655-4c5254603344 h1:vGXIOMxbNfDTk/aXCmfdLgkrSV+Z2tcbze+pEc3v5W4=
golang.org/x/net v0.0.0-20200625001655-4c5254603344/go.mod h1:/O7V0waA8r7cgGh81Ro3o1hOxt32SMVPicZroKQ2sZA=
golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d h1:TzXSXBo42m9gQenoE3b9BGiEpg5IG2JkU5FkPIawgtw=
golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d/go.mod h1:gOpvHmFTYa4IltrdGE7lF6nIHvwfUNPOp7c8zoXwtLw=
golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sync v0.0.0-20190423024810-112230192c58/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sync v0.0.0-20200625203802-6e8e738ad208/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
golang.org/x/sys v0.0.0-20190412213103-97732733099d/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd h1:xhmwyvizuTgC2qz7ZlMluP20uW+C3Rm0FD/WLDX8884=
golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
golang.org/x/text v0.3.0 h1:g61tztE5qeGQ89tm6NTjjM9VPIm088od1l6aSorWRWg=
golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
golang.org/x/tools v0.0.0-20191119224855-298f0cb1881e/go.mod h1:b+2E5dAYhXwXZwtnZ6UAqBI28+e2cm9otk0dWdXHAEo=
golang.org/x/tools v0.0.0-20200302225559-9b52d559c609 h1:3/QY44rOqJoMLCsQz9bAgInYa08qsu+dH52Uk4DWH3w=
golang.org/x/tools v0.0.0-20200302225559-9b52d559c609/go.mod h1:TB2adYChydJhpapKDTa4BR/hXlZSLoq2Wpct/0txZ28=
golang.org/x/tools v0.0.0-20200827010519-17fd2f27a9e3 h1:r3P/5xOq/dK1991B65Oy6E1fRF/2d/fSYZJ/fXGVfJc=
golang.org/x/tools v0.0.0-20200827010519-17fd2f27a9e3/go.mod h1:njjCfa9FT2d7l9Bc6FUM5FLjQPp3cFF28FI3qnDFljA=
golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898 h1:/atklqdjdhuosWIl6AIbOeHJjicWYPqR9bpxqxYG2pA=
golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543 h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=
golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
google.golang.org/appengine v1.4.0 h1:/wp5JvzpHIxhs/dumFmF7BXTf3Z+dd4uXta4kVyO508=
google.golang.org/appengine v1.4.0/go.mod h1:xpcJRLb0r/rnEns0DIKYYv+WjYCduHsrkT7/EB5XEv4=
