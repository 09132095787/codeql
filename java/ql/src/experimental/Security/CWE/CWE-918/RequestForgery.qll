import java
import semmle.code.java.frameworks.Networking
import semmle.code.java.frameworks.ApacheHttp
import semmle.code.java.frameworks.spring.Spring
import semmle.code.java.frameworks.JaxWS
import semmle.code.java.frameworks.javase.Http
import semmle.code.java.dataflow.DataFlow
private import semmle.code.java.dataflow.ExternalFlow

predicate requestForgeryStep(DataFlow::Node pred, DataFlow::Node succ) {
  // propagate to a URI when its host is assigned to
  exists(UriCreation c | c.getHostArg() = pred.asExpr() | succ.asExpr() = c)
  or
  // propagate to a URL when its host is assigned to
  exists(UrlConstructorCall c | c.getHostArg() = pred.asExpr() | succ.asExpr() = c)
  or
  // propagate to a RequestEntity when its url is assigned to
  exists(MethodAccess m |
    m.getMethod().getDeclaringType() instanceof SpringRequestEntity and
    (
      m.getMethod().hasName(["get", "post", "head", "delete", "options", "patch", "put"]) and
      m.getArgument(0) = pred.asExpr() and
      m = succ.asExpr()
      or
      m.getMethod().hasName("method") and
      m.getArgument(1) = pred.asExpr() and
      m = succ.asExpr()
    )
  )
  or
  // propagate from a `RequestEntity<>$BodyBuilder` to a `RequestEntity`
  // when the builder is tainted
  exists(MethodAccess m, RefType t |
    m.getMethod().getDeclaringType() = t and
    t.hasQualifiedName("org.springframework.http", "RequestEntity<>$BodyBuilder") and
    m.getMethod().hasName("body") and
    m.getQualifier() = pred.asExpr() and
    m = succ.asExpr()
  )
}

private class RequestForgerySinkModel extends SinkModelCsv {
  override predicate row(string row) {
    row =
      [
        "java.net;URL;false;openConnection;;;Argument[-1];request-forgery-sink",
        "java.net;URL;false;openStream;;;Argument[-1];request-forgery-sink",
        "org.apache.http.client.methods;HttpRequestBase;true;setURI;;;Argument[0];request-forgery-sink",
        "org.apache.http.message;BasicHttpRequest;true;setURI;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;HttpRequestBase;true;.ctor;;;Argument[0];request-forgery-sink",
        "org.apache.http.message;BasicHttpRequest;true;.ctor;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;setURI;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;get;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;post;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;put;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;options;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;head;;;Argument[0];request-forgery-sink",
        "org.apache.http.client.methods;RequestBuilder;false;delete;;;Argument[0];request-forgery-sink",
        "java.net.http;HttpRequest$Builder;false;uri;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;patchForObject;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;getForObject;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;getForEntity;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;execute;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;exchange;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;put;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;postForObject;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;postForLocation;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;postForEntity;;;Argument[0];request-forgery-sink",
        "org.springframework.web.client;RestTemplate;false;doExecute;;;Argument[0];request-forgery-sink",
        "javax.ws.rs.client;Client;false;target;;;Argument[0];request-forgery-sink",
        "java.net.http;HttpRequest;false;newBuilder;;;Argument[0];request-forgery-sink" // todo: this might be stricter than before. Previously the package name was not checked.
      ]
  }
}

/** A data flow sink for request forgery vulnerabilities. */
abstract class RequestForgerySink extends DataFlow::Node { }

/**
 * An argument to `org.springframework.http.RequestEntity`s constructor call
 *   which is an URI taken as a sink for request forgery vulnerabilities.
 */
private class RequestEntityUriArg extends RequestForgerySink {
  RequestEntityUriArg() {
    exists(ClassInstanceExpr e, Argument a |
      e.getConstructedType() instanceof SpringRequestEntity and
      e.getAnArgument() = a and
      a.getType() instanceof TypeUri and
      this.asExpr() = a
    )
  }
}
