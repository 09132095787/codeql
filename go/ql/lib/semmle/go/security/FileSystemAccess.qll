import go

class FastHttpFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  FastHttpFileSystemAccess() {
    exists(DataFlow::Method mcn |
      (
        mcn.hasQualifiedName("github.com/valyala/fasthttp.RequestCtx", ["SendFileBytes", "SendFile"]) or
        mcn.hasQualifiedName("github.com/valyala/fasthttp.Response", ["SendFile"])
      ) and
      this = mcn.getACall()
    )
    or
    exists(DataFlow::Function f |
      f.hasQualifiedName("github.com/valyala/fasthttp",
        [
          "ServeFile", "ServeFileUncompressed", "ServeFileBytes", "ServeFileBytesUncompressed",
          "SaveMultipartFile"
        ]) and
      this = f.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() {
    this.getTarget().getName() =
      [
        "ServeFile", "ServeFileUncompressed", "ServeFileBytes", "ServeFileBytesUncompressed",
        "SaveMultipartFile"
      ] and
    result = this.getArgument(1)
    or
    this.getTarget().getName() = ["SendFile", "SendFileBytes"] and
    result = this.getArgument(0)
  }
}

class HttpServeFile extends FileSystemAccess::Range, DataFlow::CallNode {
  HttpServeFile() {
    exists(DataFlow::Function mcn |
      mcn.hasQualifiedName("net/http", "ServeFile") and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(2) }
}

class BeegoFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  BeegoFileSystemAccess() {
    exists(DataFlow::Method mcn |
      (
        mcn.hasQualifiedName("github.com/beego/beego/v2/server/web/context.BeegoOutput", "Download") or
        mcn.hasQualifiedName("github.com/beego/beego/v2/server/web.Controller",
          "SaveToFileWithBuffer")
      ) and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() {
    this.getTarget()
        .hasQualifiedName("github.com/beego/beego/v2/server/web/context.BeegoOutput", "Download") and
    result = this.getArgument(0)
    or
    this.getTarget()
        .hasQualifiedName("github.com/beego/beego/v2/server/web.Controller", "SaveToFileWithBuffer") and
    result = this.getArgument(1)
  }
}

class EchoFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  EchoFileSystemAccess() {
    exists(DataFlow::Method mcn |
      mcn.hasQualifiedName("github.com/labstack/echo/v4.Context", ["Attachment", "File"]) and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(0) }
}

class GinFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  GinFileSystemAccess() {
    exists(DataFlow::Method mcn |
      mcn.hasQualifiedName("github.com/gin-gonic/gin.Context",
        ["File", "FileAttachment", "SaveUploadedFile"]) and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() {
    this.getTarget().getName() = ["File", "FileAttachment"] and result = this.getArgument(0)
    or
    this.getTarget().getName() = "SaveUploadedFile" and result = this.getArgument(1)
  }
}

class IrisFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  IrisFileSystemAccess() {
    exists(DataFlow::Method mcn |
      mcn.hasQualifiedName("github.com/kataras/iris/v12/context.Context",
        ["SendFile", "ServeFile", "SendFileWithRate", "ServeFileWithRate", "UploadFormFiles"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/kataras/iris/v12/context.Context", "SaveFormFile") and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() {
    this.getTarget().getName() =
      ["SendFile", "ServeFile", "SendFileWithRate", "ServeFileWithRate", "UploadFormFiles"] and
    result = this.getArgument(0)
    or
    this.getTarget().getName() = "SaveFormFile" and result = this.getArgument(1)
  }
}

class FiberSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  FiberSystemAccess() {
    exists(DataFlow::Method mcn |
      mcn.hasQualifiedName("github.com/gofiber/fiber/v2.Ctx", ["Attachment", "SendFile"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/gofiber/fiber/v2.Ctx", "SaveFile") and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() {
    this.getTarget().getName() = ["Attachment", "SendFile"] and result = this.getArgument(0)
    or
    this.getTarget().getName() = "SaveFile" and result = this.getArgument(1)
  }
}

class AferoSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  AferoSystemAccess() {
    exists(DataFlow::Function mcn |
      mcn.hasQualifiedName("github.com/spf13/afero",
        ["WriteReader", "SafeWriteReader", "WriteFile", "ReadFile", "ReadDir"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero", "Open") and
      this = mcn.getACall()
    )
    or
    exists(DataFlow::Function mcn |
      mcn.hasQualifiedName("github.com/spf13/afero",
        ["WriteReader", "SafeWriteReader", "WriteFile", "ReadFile", "ReadDir"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero", "Open") and
      this = mcn.getACall()
    )
    or
    exists(DataFlow::Method mcn |
      mcn.hasQualifiedName("github.com/spf13/afero.Afero",
        ["ReadFile", "ReadDir", "WriteReader", "WriteFile", "SafeWriteReader"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.HttpFs", ["Open", "OpenFile", "Create"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.RegexpFs",
        ["Create", "Open", "Remove", "OpenFile"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.ReadOnlyFs",
        ["Create", "Open", "Remove", "OpenFile", "ReadDir", "ReadlinkIfPossible"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.OsFs",
        ["Create", "Open", "Remove", "RemoveAll", "OpenFile", "ReadDir", "ReadlinkIfPossible"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.OsFs",
        ["Create", "Open", "Remove", "RemoveAll", "OpenFile", "ReadDir", "ReadlinkIfPossible"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.MemMapFs",
        ["Create", "Open", "OpenFile", "Remove", "RemoveAll"]) and
      this = mcn.getACall()
      or
      mcn.hasQualifiedName("github.com/spf13/afero.BasePathFs",
        ["Create", "Open", "OpenFile", "Remove", "RemoveAll", "ReadlinkIfPossible"]) and
      this = mcn.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() {
    if
      this.getTarget()
          .hasQualifiedName("github.com/spf13/afero",
            ["WriteReader", "SafeWriteReader", "WriteFile", "ReadFile", "ReadDir"]) or
      this.getTarget()
          .hasQualifiedName("github.com/spf13/afero",
            ["WriteReader", "SafeWriteReader", "WriteFile", "ReadFile", "ReadDir"])
    then result = this.getArgument(1)
    else result = this.getArgument(0)
  }
}
