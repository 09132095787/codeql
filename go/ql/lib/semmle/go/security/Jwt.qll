/**
 * Provides classes and predicates for reasoning about JSON Web Tokens (JWT).
 */

import go
private import semmle.go.security.HardcodedCredentials

private class IrisJwt extends HardcodedCredentials::Sink {
  IrisJwt() {
    exists(Field f |
      f.hasQualifiedName(package("github.com/kataras/iris", "middleware/jwt"), "Signer", "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}

private class GogfJwtSign extends HardcodedCredentials::Sink {
  GogfJwtSign() {
    exists(Field f |
      f.hasQualifiedName(package("github.com/gogf/gf-jwt", ""), "GfJWTMiddleware", "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}

private class GinJwtSign extends HardcodedCredentials::Sink {
  GinJwtSign() {
    exists(Field f |
      f.hasQualifiedName(package("github.com/appleboy/gin-jwt", ""), "GinJWTMiddleware", "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}

private class SquareJoseKey extends HardcodedCredentials::Sink {
  SquareJoseKey() {
    exists(Field f, string pkg |
      pkg = ["github.com/square/go-jose/v3", "gopkg.in/square/go-jose.v2"]
    |
      f.hasQualifiedName(pkg, ["Recipient", "SigningKey"], "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}
