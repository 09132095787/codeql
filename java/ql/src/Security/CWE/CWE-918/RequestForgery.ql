/**
 * @name Server-side request forgery
 * @description Making web requests based on unvalidated user-input
 *              may cause the server to communicate with malicious servers.
 * @kind path-problem
 * @problem.severity error
 * @precision high
 * @id java/ssrf
 * @tags security
 *       external/cwe/cwe-918
 */

import java
import semmle.code.java.dataflow.FlowSources
import semmle.code.java.security.RequestForgery
import DataFlow::PathGraph

class RequestForgeryConfiguration extends TaintTracking::Configuration {
  RequestForgeryConfiguration() { this = "Server-Side Request Forgery" }

  override predicate isSource(DataFlow::Node source) {
    source instanceof RemoteFlowSource and
    // Exclude results of remote HTTP requests: fetching something else based on that result
    // is no worse than following a redirect returned by the remote server, and typically
    // we're requesting a resource via https which we trust to only send us to safe URLs.
    not source.asExpr().(MethodAccess).getCallee() instanceof URLConnectionGetInputStreamMethod
  }

  override predicate isSink(DataFlow::Node sink) { sink instanceof RequestForgerySink }

  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
    any(RequestForgeryAdditionalTaintStep r).propagatesTaint(pred, succ)
  }

  override predicate isSanitizer(DataFlow::Node node) { node instanceof RequestForgerySanitizer }
}

from DataFlow::PathNode source, DataFlow::PathNode sink, RequestForgeryConfiguration conf
where conf.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "Potential server-side request forgery due to $@.",
  source.getNode(), "a user-provided value"
