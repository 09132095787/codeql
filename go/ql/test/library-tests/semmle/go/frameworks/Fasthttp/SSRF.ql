import go
import semmle.go.security.RequestForgery

select any(RequestForgery::Sink s)
