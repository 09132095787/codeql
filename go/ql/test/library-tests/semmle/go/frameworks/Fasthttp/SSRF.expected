| fasthttp.go:14:42:14:57 | "127.0.0.1:8909" |
| fasthttp.go:15:32:15:46 | "google.com:80" |
| fasthttp.go:16:39:16:53 | "google.com:80" |
| fasthttp.go:17:48:17:62 | "google.com:80" |
| fasthttp.go:28:24:28:46 | "http://127.0.0.1:8909" |
| fasthttp.go:29:32:29:54 | "http://127.0.0.1:8909" |
| fasthttp.go:30:31:30:53 | "http://127.0.0.1:8909" |
| fasthttp.go:31:25:31:47 | "http://127.0.0.1:8909" |
| fasthttp.go:33:14:33:16 | req |
| fasthttp.go:34:23:34:25 | req |
| fasthttp.go:35:22:35:24 | req |
| fasthttp.go:36:21:36:23 | req |
| fasthttp.go:55:26:55:48 | "http://127.0.0.1:8909" |
| fasthttp.go:56:34:56:56 | "http://127.0.0.1:8909" |
| fasthttp.go:57:33:57:55 | "http://127.0.0.1:8909" |
| fasthttp.go:58:27:58:49 | "http://127.0.0.1:8909" |
| fasthttp.go:59:16:59:18 | req |
| fasthttp.go:60:24:60:26 | req |
| fasthttp.go:61:25:61:27 | req |
| fasthttp.go:62:23:62:25 | req |
| fasthttp.go:66:14:66:16 | req |
| fasthttp.go:67:22:67:24 | req |
| fasthttp.go:68:21:68:23 | req |
| fasthttp.go:71:13:71:19 | resByte |
| fasthttp.go:72:21:72:27 | resByte |
| fasthttp.go:73:20:73:26 | resByte |
| fasthttp.go:74:14:74:20 | resByte |
| fasthttp.go:75:12:75:14 | req |
| fasthttp.go:76:20:76:22 | req |
| fasthttp.go:77:21:77:23 | req |
| fasthttp.go:78:19:78:21 | req |
| fasthttp.go:81:20:81:22 | req |
| fasthttp.go:82:28:82:30 | req |
| fasthttp.go:83:27:83:29 | req |
| fasthttp.go:86:17:86:32 | "127.0.0.1:8909" |
| fasthttp.go:87:24:87:39 | "127.0.0.1:8909" |
| fasthttp.go:88:26:88:41 | "127.0.0.1:8909" |
| fasthttp.go:89:33:89:48 | "127.0.0.1:8909" |
